### PR TITLE
1-add test for product name validation

### PR DESCRIPTION
**Pull request by ebarylko**
## Summary

Added the `Issue` class and the functions that operate on it.

Added a test for `validateProductName`.

### src/main/kotlin/Issues.kt
* Added types: `Issue`, `Description`, `Priority`, `IssueInformation`, `IssueId`, `Days`, `Status`, `IssueFilter`, `IssuePage`, `RequestPage` 
* Added functions: `searchIssues`, `nextPage`, `searchIssues`, `editIssue`, `listRequests`, `issuesMenu`, `enterIssueInformation`, `saveIssue`

### src/main/kotlin/Product.kt
* Added `validateProductName`
* Added `Product` which represents a valid product type

### src/main/kotlin/Release.kt
* Added `ReleaseId` which represents a valid release id type

### src/main/kotlin/Request.kt
* Added `Request` which represents a valid request type

### src/main/kotlin/Contact.kt
* Added `Contact` which represents a valid contact type

### src/test/kotlin/ProductTest.kt
* Added a test for `validateProductName`

### CODE REVIEW

File: src/main/kotlin/Contact.kt [hunk 1]
```diff
@@ -0,0 +1,44 @@
+@JvmInline
```
Comment (micahdbak) [inline src/main/kotlin/Contact.kt:1]: I'm noticing very quickly that these simple definitions span many lines. Maybe we can adjust ktlint to make these files a little shorter?

It makes the files less approachable in my opinion.

File: src/main/kotlin/Contact.kt [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    val email: String,
+) {
+    init {
+        require(email.contains("@") && email.length in 5..24) {
+            "Email does not contain @ or is not within 1 to 30 characters"
+        }
+    }
+}
+
+@JvmInline
```
Comment (micahdbak) [inline src/main/kotlin/Contact.kt:23]: What does `@JvmInline` do? Is it necessary + something we should explain in our appendix?

File: src/main/kotlin/Issues.kt [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+    init {
+        require(id in 1..<100) {
+            "Id must be a positive integer within [1, 99]"
+        }
+    }
+}
+
+data class Issue(
+    val id: IssueId,
+    val information: IssueInformation,
```
Comment (micahdbak) [inline src/main/kotlin/Issues.kt:64]: Is there a reason the contents of `IssueInformation` aren't directly in `Issue`? This seems like unnecessary nesting

File: src/main/kotlin/Issues.kt [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+    val createdAt: TimeSource,
```
Comment (micahdbak) [inline src/main/kotlin/Issues.kt:65]: Is there a simpler way we can store the date without leaning on something like this? We may need to do some funky conversions between the DB and this type, and we will only be taking in strings of the form YYYY-MM-DD from the user. Maybe we can specify our own type like:

```kotlin
data class Date(
  val year: Int,
  val month: Int,
  val day: Int
)
```

File: src/main/kotlin/Issues.kt [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+    init {
+        require(numOfDays >= 0) {
+            "The number of days must be non-negative"
+        }
+    }
+}
+
+sealed class Status {
+    data object Assessed : Status()
+
```
Comment (micahdbak) [inline src/main/kotlin/Issues.kt:81]: Re: my above comment on ktlint, are these blank lines necessary? they're sort of extra.

File: src/main/kotlin/Issues.kt [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+    data object Created : Status()
+
+    data object Done : Status()
+
+    data object Cancelled : Status()
+
+    data object InProgress : Status()
+}
+
+sealed class IssueFilter {
```
Comment (micahdbak) [inline src/main/kotlin/Issues.kt:91]: I'm wondering if the contents of `IssueFilter` are going a bit far. Kotlin may be blurring the lines between what are skeleton classes/functions on the basis of compile-ability, but this feels more like implementation than "header" logic.

File: src/main/kotlin/Issues.kt [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+    data class ByProduct(
+        val product: Product,
+    ) : IssueFilter()
+
+    data class Composite(
+        val filter: List<IssueFilter>,
+    ) : IssueFilter()
+}
+
+data class PageOf<T>(
```
Comment (micahdbak) [inline src/main/kotlin/Issues.kt:121]: Ditto for this

File: src/main/kotlin/Main.kt [hunk 8]
```diff
@@ -1,141 +0,0 @@
-import com.varabyte.kotter.foundation.*
```
Comment (micahdbak) [inline src/main/kotlin/Main.kt:1]: I'll open a PR adding my Main.kt file to this repo

**Review by micahdbak — COMMENTED** [review]
Looks pretty good so far; my primary concerns are about thoroughness. I think we can make the code more "skeleton"y

File: src/main/kotlin/Contact.kt (hunk 2 above)
Comment (ebarylko) [reply src/main/kotlin/Contact.kt:23]: Value classes can only be created when using the `@JvmInline` annotation, which specifies that the value class is also an inline class.

Comment (ebarylko) [reply src/main/kotlin/Contact.kt:23]: Having it in the appendix would make it easier for the TA to understand why it is present. Good catch!

Comment (micahdbak) [reply src/main/kotlin/Contact.kt:23]: Interesting, and what makes a value class different from a regular class? Or a data class?

Comment (micahdbak) [reply src/main/kotlin/Contact.kt:23]: Forgive me as I don't have much Kotlin experience 🤣

File: src/main/kotlin/Issues.kt (hunk 4 above)
Comment (ebarylko) [reply src/main/kotlin/Issues.kt:65]: I understand your concern about the conversion of types to and from the DB, but I am confused about your comment about taking dates from the user. I thought that the only time we generate dates are when issues are created. Are there other cases where we generate a date?

Comment (micahdbak) [reply src/main/kotlin/Issues.kt:65]: Oh that's a good point. I was thinking back to our requirements specification where I created BNFs for the dates - but I think you're right, none of our use cases actually get the date from users. I'll keep this in mind for when we update the req spec. then. I have no problem with your use of the TimeSource type then :)

File: src/main/kotlin/Contact.kt (hunk 1 above)
Comment (ebarylko) [reply src/main/kotlin/Contact.kt:1]: Ktlint does allow you to create custom configuration for it to follow.

What did you have in mind for making the definitions shorter?

File: src/main/kotlin/Issues.kt (hunk 5 above)
Comment (ebarylko) [reply src/main/kotlin/Issues.kt:81]: Ok. I will remove the blank lines.

File: src/main/kotlin/Contact.kt (hunk 1 above)
Comment (micahdbak) [reply src/main/kotlin/Contact.kt:1]: I'll need to look into ktlint so I'll get back to you on this - but good to know!

File: src/main/kotlin/Issues.kt (hunk 6 above)
Comment (ebarylko) [reply src/main/kotlin/Issues.kt:91]: Could you please clarify more on what you mean?

File: src/main/kotlin/Contact.kt (hunk 2 above)
Comment (ebarylko) [reply src/main/kotlin/Contact.kt:23]: A value class is a way to represent a single value as a type within the domain, retaining the optimization benefits associated with the specific type of the value. If you had encapsulated that value within a regular class or data class, you would not have the optimization benefits as if you were working directly with the type.
https://kotlinlang.org/docs/inline-classes.html

**Comment by ebarylko** [comment]
All the work from this PR has already been merged.

**Pull request closed without merging**